contributor-guide/modules/ROOT/pages/testing/fuzzing.adoc

```diff
@@ -0,0 +1,174 @@
+= Fuzz testing
+
+*What is fuzz testing?* Fuzzing is a testing technique that injects random pieces of data to a software function to uncover crashes and vulnerabilities. It helps improving code security and reliability, since it can trigger edge cases that went unnoticed during unit testing.
+
+*Should I use it?* Fuzz testing is specially relevant for libraries that process potentially untrusted, user-controlled input, like network data. Libraries that implement parsers, decoders or network protocols usually benefit from fuzz testing.
+
+*Which Boost libraries use it?* Libraries like boost:json[], boost:url[] and boost:mysql[] use this technique - if you're about to implement it in your library, have a look at what these libraries do.
+
+*Should I still write unit tests?* Yes. Absolutely. Fuzzing does not replace unit tests, but complements them. Unit tests verify that your code produces the intended results by providing known inputs and running assertions on the outputs. In fuzz testing, inputs are generated randomly by the fuzzing engine, so no assertions are usually run on the outputs - fuzzing will only monitor for crashes and memory errors.
+
+*How can I add fuzzing to my library?* We recommend using https://llvm.org/docs/LibFuzzer.html[LibFuzzer], since it's the easiest fuzzing engine to use, and the one that other Boost libraries use. You can use other fuzzing engines if you prefer.
+
+== LibFuzzer basics
+
+Quoting documentation, "LibFuzzer is an in-process, coverage-guided, evolutionary fuzzing engine". LibFuzzer will run your code multiple times with different, random inputs. It will instrument your code to measure coverage, and will attempt to generate inputs that maximize it, effectively trying to discover new paths in your code.
+
+LibFuzzer is included in `clang`, so you don't need to install anything to get started.
+
+Let's say we want to fuzz a function that parses JSON data, like `parse_json(string_view input)`. We will create a source file with the following code:
+
+[source,cpp]
+----
+
+#include <string_view>
+#include <your/parsing/function.hpp>
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size)
+{
+    // The range [data, data+size) contains the data generated by the fuzzer
+    std::string_view input_data (reinterpret_cast<const char*>(data), size);
+    parse_json(input_data);
+    return 0;
+}
+
+----
+
+
+We can build a fuzzer executable by adding `-fsanitize=fuzzer` to clang's compile and link flags. This will automatically link LibFuzzer to your code. It's advised to also enable the address and undefined sanitizers, which increases the range of errors detected by the fuzzer. We recommend building in release mode with debug symbols enabled, so crashes are symbolized correctly.
+
+From the command line:
+
+[source,bash]
+----
+clang++ -g -O3 -fsanitize=fuzzer,address,undefined -o fuzzer fuzzer.cpp 
+----
+
+Or as a `Jamfile` target:
```

Comment:
Maybe also CMake? I'm not sure.

Reply:
Should release mode be capitalized? 
You say look at the libraries URL, MySQl, Json - is there a link we can add that shows the fuzz testing functions added - in which case perhaps mention the names of functions to look out for?

Reply:
Can you give an example, or perhaps link to, a sample (or two) Corpus?
Perhaps also add links to the Continuous Integration topic, Writing Tests, and any other closely related documentation. Add the link on first mention in the text, if it is mentioned, or in a See Also section at the end - if the topic is related but not explicitly mentioned.
Great content - the value of a topic like this is educating the reader into what are the "Next Steps" to get this going - which I think you have addressed well.

Reply:
All fuzzers in Boost are currently run from b2. The only issue with cmake is that you need to specify an option to prevent fuzzing from building by default - while in b2 you just don't `build-project fuzzing` and go. Since it may help someone, I've added it tho.



That said, if you think it's is valuable, I can add it.